### ISSUE TITLE
[MASSEMBLY-6] dependency management in AbstractUnpackingMojo is broken in the reactor

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-assembly-plugin#305]
**[Jerome Lacoste](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lacostej)** opened **[MASSEMBLY-6](https://issues.apache.org/jira/browse/MASSEMBLY-6?redirect=false)** and commented

This problem only affects reactor builds.

MNG-1206 broke my build. The lines

    String key = artifact.getGroupId() + ":" + artifact.getArtifactId() + ":" + artifact.getVersion();
    if ( !reactorArtifacts.containsKey( key ) && !dependencies.containsKey( key ) )
    {
        dependencies.put( key, artifact );
    }

Any dependency that is supposedly known by the reactor won't be included in the project's dependencies.

So if you have:

moduleA
moduleB depends on moduleA

moduleA will be missing in the dependencies found by getDependencies() in an assembly plugin used by moduleB, as it is already in the reactorArtifacts. Not good when you want to use moduleA in your assembly :)

Also note that the key used doesn't take into account the type of the artifact. That may be a problem on its own, perhaps with attached artifacts.

Shouldn't getDependencies() just work in the context of the reactor instead of this particular mojo having to identify the relevant dependencies?
One thing is to fix that properly. In the mean time, I am leaning on removing the check (!reactorArtifacts.containsKey( key )).
I'd rather have too many dependencies than not enough. Q: would this affect the UnpackMojo ?

Related issue: I also find strange that we process all reactor dependencies. I think this could lead to the situation where a dependency not referenced by the pom but referenced by the assembly config file and referenced by a module in the reactor which has not yet been built could be taken into account. We should restrict the dependencies from the reactorProjects up to the current project being built.

Basically:

moduleA depends on lib1
moduleB depends on lib2

When we are using the assembly in moduleA from the reactor, lib2 shouldn't be in the list of dependencies. Today I think it is.

All in all I think the getDependencies() is not correct. I will try removing the reactorArtifacts check for now, but it clearly could be improved.

We shouldn't even add dependencies of all reactorProjets. Those after the current project should be discarded

---

**Attachments:**
- [MNG-1588.diff](https://issues.apache.org/jira/secure/attachment/12714064/MNG-1588.diff) (_2.48 kB_)

**Comment by jira-importer** [comment]
**[Jerome Lacoste](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lacostej)** commented

This fixes the issue, althought I believe it to not be the full correct fix.
Once this is checked in, the priority can be downgraded from blocker to normal (or we create a separate issue for the remaining cleanup).

**Comment by jira-importer** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

applied, with changes to the comments.
